src/aiida_qe_xspec/calculations/functions/xspectra/get_xps_spectra.py

```diff
@@ -3,109 +3,105 @@
                 final_spectra_y_arrays.append(
-                    intensity * voigt_profile(x_energy_range - relative_peak_position, sigma, gamma) /
-                    total_multiplicity
+                    intensity* voigt_profile(x_energy_range - relative_peak_position, sigma, gamma)
                 )
-
-            final_spectra_y_labels.append(f'{element}_total_xps')
+            final_spectra_y_labels.append(f'{element}_total')
             final_spectra_y_units.append('sigma')
             final_spectra_y_arrays.append(sum(final_spectra_y_arrays))
-
             final_spectra_x_label = 'energy'
             final_spectra_x_units = 'eV'
             final_spectra_x_array = x_energy_range
             final_spectra.set_x(final_spectra_x_array, final_spectra_x_label, final_spectra_x_units)
             final_spectra.set_y(final_spectra_y_arrays, final_spectra_y_labels, final_spectra_y_units)
-            result_spectra[f'{element}_{label}'] = final_spectra
-        return result_spectra
+            result_spectra[f'{element}_{orbital}'] = final_spectra
+    return result_spectra
+
+@calcfunction
+def get_spectra_by_element(core_levels, equivalent_sites_data, voight_gamma, voight_sigma, **kwargs):  # pylint: disable=too-many-statements
+    """Generate the XPS spectra for each element.
+
+    Calculate the core level shift and binding energy for each element.
+    Generate the final spectra using the Voigt profile.
+
+    :param core_levels: a Dict object defining the elements and their core-levels to consider
+            when producing spectra, e.g., {"C": ["1s"], "Al": ["2s", "2p"]}.
+    :param equivalent_sites_data: an Dict object containing symmetry data.
+    :param voight_gamma: a Float node for the gamma parameter of the voigt profile.
+    :param voight_sigma: a Float node for the sigma parameter of the voigt profile.
+    :param structure: the StructureData object to be analysed
+    :returns: Dict objects for all generated spectra and associated binding energy
+            and core level shift.
+
+    """
+    from copy import deepcopy
 
-    result.update(spectra_broadening(core_level_shifts))
+    ground_state_node = kwargs.pop('ground_state', None)
+    correction_energies = kwargs.pop('correction_energies', orm.Dict()).get_dict()
+    ch_nodes = kwargs.pop('ch_nodes', {})
+    group_state_energy = ground_state_node.get_dict()['energy'] if ground_state_node is not None else None
+    core_levels = core_levels.get_dict()
+    equivalency_data = equivalent_sites_data.get_dict()
+
+    data_dict = {element: {} for element in core_levels.keys()}
+    for key, xspectra_out_params in ch_nodes.items():
+        element, _, site, orbital = key.split('_')
+        data_dict[element].setdefault(orbital, {})
+        energy = xspectra_out_params.get_dict()['energy']
+        data_dict[element][orbital][f'site_{site}'] = {'energy': energy, 'multiplicity': equivalency_data[f'site_{site}']['multiplicity']}
```

Comment:
As this section depends on a specific formatting of the keys for each `ch_node` (e.g. `c_site_1_1s`), *any* deviation from that due to the core level string in `core_levels`, `pseudos`, or `core_level_pseudos` being written incorrectly (e.g. `1_s`, instead of `1s`) needs to be checked before the `XpsWorkChain` is started - otherwise the process will except only at the end of the run.

I would suggest adding input validation for the `XpsWorkChain` to ensure that `key.split()` does actually produce something like `['c', 'site', '1', '1s']` and return an error if not.

Reply:
Thanks for pointing out this. I have revised the code to use the exact keys (element, orbital, site), instead of using a `string` and `split`., please check the commit [f2d6e3f](https://github.com/aiidaplugins/aiida-qe-xspec/pull/16/commits/f2d6e3fa353e035780e3dc56c14d213b1414b8d6)